internals/overlord/state/identities.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+// TODO: disallow multiple users with same "local: user-id" value?
+
+// Identity holds the configuration of a single identity.
+type Identity struct {
+	Name   string
+	Access IdentityAccess
+
+	// One or more of the following type-specific configuration fields must be
+	// non-nil (currently the only type is "local").
+	Local *LocalIdentity
+}
+
+// IdentityAccess defines the access level for an identity.
+type IdentityAccess string
+
+const (
+	AdminAccess     IdentityAccess = "admin"
+	ReadAccess      IdentityAccess = "read"
+	UntrustedAccess IdentityAccess = "untrusted"
+)
+
+// LocalIdentity holds identity configuration specific to the "local" type
+// (for ucrednet/UID authentication).
+type LocalIdentity struct {
+	UserID uint32
+}
+
+// validate checks that identity d is valid, returning an error if not.
+func (d *Identity) validate() error {
+	if d == nil {
+		return errors.New("identity must not be nil")
+	}
+
+	switch d.Access {
+	case AdminAccess, ReadAccess, UntrustedAccess:
+	default:
+		return fmt.Errorf("invalid access %q, must be %q, %q, or %q",
+			d.Access, AdminAccess, ReadAccess, UntrustedAccess)
+	}
+
+	switch {
+	case d.Local != nil:
+		if d.Local.UserID == 0 {
```

Comment:
I understand this for root, makes sense (go int default value and root uid sharing meaning). For the pebble user (the user pebble daemon is running as), I think it is a bit different and we have to be softer there, as someone may change the uid the daemon is running as.

So if there is an identity that matches the pebble user uid, I think we just ignore the value of Access as it is always "AdminAccess".

Reply:
My comment about making UserID a string might make sense, so we can technically distinguish between empty and 0, without resorting to uint32 ptr.

Reply:
I've updated this to use `*uint32` to allow us to distinguish between 0 (now valid) and unset (invalid) at the JSON/YAML level.

I believe with how I've implemented `AdminAccess.CheckAccess` and `UserAccess.CheckAccess` (see access.go), they fall back to looking at ucred directly if there wasn't an identity found. Does that seem reasonable?

Reply:
Yep. Thank-you.